### PR TITLE
Support test discovery and execution as first class features

### PR DESCRIPTION
**Pull request by rafaelbey**
This creates two new JRPC's to better support the testable discovery and executions thru VS Code TestController.

https://code.visualstudio.com/api/extension-guides/testing

NOTE - I did not forgot about legacy tests (mapping, service), but for now decided to leaves those untouched (they will go thru commands).

sample testing

### CODE REVIEW

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java [hunk 1]
```diff
@@ -71,79 +62,122 @@ public final class TestableCommandsSupport implements CommandsSupport
...
-    public Iterable<? extends LegendExecutionResult> executeCommand(SectionState section, PackageableElement element, String commandId, Map<String, String> executableArgs, Map<String, Object> inputParameters)
+    public List<LegendTestExecutionResult> executeTests(SectionState sectionState, PackageableElement element, String testId, Set<String> excludedTestIds)
     {
-        TextLocation location = SourceInformationUtil.toLocation(element.sourceInformation);
-        String entityPath = element.getPath();
-        switch (commandId)
+        return this.collectTests(element)
+                .map(test ->
+                {
+                    LinkedList<LegendTest> q = new LinkedList<>();
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java:109]: I think ArrayDeque is a better queue implementation than LinkedList.
```suggestion
                    Deque<LegendTest> q = new ArrayDeque<>();
```

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java [hunk 2]
```diff
@@ -71,79 +62,122 @@ public final class TestableCommandsSupport implements CommandsSupport
...
-            }
-            case RUN_TEST_COMMAND_ID:
+        return this.executeTests(sectionState, element, uniqueTestIds);
+    }
+
+    private List<UniqueTestId> collectUniqueTestId(LegendTest test, Set<String> excludedTestIds)
+    {
+        List<UniqueTestId> uniqueTestIds = new ArrayList<>();
+
+        LinkedList<LegendTest> q = new LinkedList<>();
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java:157]:
```suggestion
        Deque<LegendTest> q = new ArrayDeque<>();
```

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        if (this == o)
+        {
+            return true;
+        }
+        if (!(o instanceof LegendTest))
+        {
+            return false;
+        }
+        LegendTest that = (LegendTest) o;
+        return Objects.equals(id, that.id);
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:87]: I don't think it's possible for id to be null.
```suggestion
        return this.id.equals(that.id);
```

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(id);
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:93]: I don't think it's possible for id to be null.
```suggestion
        return this.id.hashCode();
```

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    private final String id;
+    private final String label;
+    private final List<LegendTest> children;
+    private final TextLocation location;
+
+    private LegendTest(List<String> idComponents, List<LegendTest> children, TextLocation location)
+    {
+        this.idComponents = idComponents;
+        this.id = String.join(".", idComponents);
+        this.label = idComponents.get(idComponents.size() - 1);
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:36]: What if idComponents is empty? Is that possible?

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+{
+    private final List<String> idComponents;
+    private final String id;
+    private final String label;
+    private final List<LegendTest> children;
+    private final TextLocation location;
+
+    private LegendTest(List<String> idComponents, List<LegendTest> children, TextLocation location)
+    {
+        this.idComponents = idComponents;
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:34]: Should this be made unmodifiable?

File: legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/LegendLanguageService.java [hunk 7]
```diff
@@ -107,4 +113,41 @@ public CompletableFuture<String> replClasspath()
...
+            this.server.getGlobalState().forEachDocumentState(docState ->
+            {
+                docState.forEachSectionState(sectionState ->
+                {
+                    LegendLSPGrammarExtension extension = sectionState.getExtension();
+                    if (extension == null)
+                    {
+                        return;
+                    }
+                    commands.addAll(extension.testCases(sectionState));
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/LegendLanguageService.java:133]:
```suggestion
                    if (extension != null)
                    {
                        commands.addAll(extension.testCases(sectionState));
                    }
```

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java (hunk 1 above)
Comment (rafaelbey) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java:109]: Fixed!

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java (hunk 2 above)
Comment (rafaelbey) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/TestableCommandsSupport.java:157]: Fixed!

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java (hunk 3 above)
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:87]: This was intellij code generation... fixed on last commit

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java (hunk 5 above)
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:36]: It should never be empty.  Adding a check on constructor

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java (hunk 6 above)
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/test/LegendTest.java:34]: YEs... done on last commit

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**